### ISSUE TITLE
Changes to js file included by require() not reflect when refresh the browser

### ISSUE THREAD

**Issue by superbull** [browserify-rails/browserify-rails#101]
I have below file structure:

```
- app
  - assets
    - javascripts
      - application.js
      - foo.js
```

``` javascript
// application.js

//= require jquery
//= require jquery_ujs
//= require turbolinks

var foo = require('./foo');
console.log(foo(10));
```

``` javascript
// foo.js

module.exports = function (n) { return n * 10 }
```

This works fine at the first time. The console will output:

```
100
```

Then if I modify the file `foo.js` to

``` javascript
// modified foo.js

module.exports = function (n) { return n * 20 }
```

and refresh the browser, the output of the console is the same to the previous one.

I can make the new code of `foo.js` take effects by doing one of the folloing:
- clear the cache
  
  ``` bash
  rake tmp:cache:clear
  ```
- modify `application.js` file, e.g. add a new line, and save the file.
- add `//= require_tree .` to the `application.js` file, this has a side effect that the `foo.js` file will be included twice.

It seems the problem is that the changes to the `foo.js` file is not recognized by the asset pipeline. Is there any settings in browserify-rails to fix this problem?

**Comment by cymen** [comment]
Have you tried this with a vanilla Rails install with all the defaults? Are you mounting any file systems via NFS to a vagrant image or anything odd like that?

**Comment by superbull** [comment]
The problem is happened on my windows 7 PC.

I have just tried it on my MAC, there is no problem. Strange!

**Comment by cymen** [comment]
That is odd. But I have no time for anything like that -- PRs welcome if you figure out a solution!

**Comment by haydenzone** [comment]
Running into the same issue on OSX Yosemite 10.10.5.

**Comment by cymen** [comment]
@haydenzone Are virtual machines involved at all (docker, vagrant, etc)?

If not, have you tried a different editor (as odd as it sounds, some editors use different OS APIs and some signal change notification better than others -- to the point that some editors have additional configuration).

**Comment by drewhamlett** [comment]
Same issue over here.

application.js
require('./setup')

setup.js
console.log('ddfg')

Only reflects changes when I change something on application.js

**Comment by cymen** [comment]
@drewhamlett Are virtual machines involved at all (docker, vagrant, etc)?

**Comment by cymen** [comment]
@drewhamlett Also are you using browserify-rails 2.0.1?

**Comment by cymen** [comment]
I am going to try to reproduce this with a fresh rails project. With 2.x, we no longer depend on tilt however we may have introduced this bug with that change. But I don't know yet.

**Comment by cymen** [comment]
I am able to reproduce. I think sprockets is much more aggressive with caching in Rails 4 and we need to inform it that it's cache also depends on the files outside of sprockets that are getting pulled in by browserify. I opened up an issue on it here:

https://github.com/rails/sprockets/issues/163

@drewhamlett @haydenzone If either of you are starting new projects without a lot of legacy JavaScript, I highly suggest you get off of the Rails asset pipeline. In other words, I advocate for not using this gem but instead using browserify or webpack directly (or via an intermediary build tool like gulp). I contributed to this gem so the company I was working for could do that. But if you want first-class JavaScript development, the asset pipeline is not a good place to be.

**Comment by drewhamlett** [comment]
@cymen Yea, looks like I'll have to join the masses of comparing Gulpfile line count and upping my e-peen.  Amazing how front end development has gotten harder over the past year.  Since I've written this sentence another javascript build tool came out.  Maybe it'll be good.

Babel used to easy to use, now everything is broken up into a thousand different modules.  NPM is the slowest thing I've ever used, and people are going to shittier and shittier front end frameworks.  I guess people don't like elegant solutions like Vue or Aurelia.  They'd rather spend time writing build scripts, doing stuff from scratch, and making their lives harder.

**Comment by cymen** [comment]
@drewhamlett I personally am using webpack direct (no Gulpfile, no anything else) with React for views/components and working towards redux for state management. I find this fairly simple, straight forward and productive. Styles are a bit tricky however my pragmatic approach the moment is to use the extract-text-webpack-plugin module -- that lets me simple do `import './style.sass` and when webpack builds the JS, extract-text-webpack-plugin extracts the CSS and it can go in a separate bundle.css. It works really well. I also use babel but as a plugin to webpack. Here is what my webpack configuration looks like:

https://gist.github.com/cymen/9aec78c717c352ea0dc2

In reality, I actually have two more webpack configurations there for building some other things so some things are extracted to variables but it should be fairly easy to follow.

**Comment by haydenzone** [comment]
Are virtual machines involved at all (docker, vagrant, etc)? Nope
 have you tried a different editor? Yup. I made a completely new rails app from command line and only used vim to modified files, and the issue still reproduced.
Also are you using browserify-rails 2.0.1? I am 

I actually ended up migrating over to https://larsjung.de/modulejs/ and have had good results. I was working on a fresh project though and didn't need all the bells and whistles of browserify.

**Comment by cymen** [comment]
@haydenzone Thanks! I was able to reproduce and hopefully we'll get a fix soon but I'm glad you're not blocked.

**Comment by cymen** [comment]
For now, I would recommend downgrading to the latest 1.x release which will resolve the issue while the cause of the breakage in 2.x is determined.

**Comment by cymen** [comment]
I found the bug -- working on a fix and will release a new version shortly.

**Comment by cymen** [comment]
@haydenzone @drewhamlett @superbull There was indeed a bug in how dependencies were being passed back to sprockets. The bug is now fixed with 2.0.2. If you are interested, please upgrade and try 2.0.2. My apologies for missing this. Thanks!

**Comment by cymen** [comment]
Also I have retracted 2.0.0 and 2.0.1 which exhibited this bug.

**Comment by codeurge** [comment]
It was noted in another related issue, and thought I'd put it here for clear visibility:

You may need to clear your browserify-inc cache by either executing `rake tmp:cache:clear`, or completely removing the cache directory - `rm -rf tmp/cache` after upgrading in order to see changes being reflected.

**Comment by gauravtiwari** [comment]
@cymen Looks like this issue still exists - tried clearing cache and restarting servers several times and still the same issue. 1.5.0 does work fine though.

**Comment by cymen** [comment]
@gauravtiwari Can you share an example repo or at least the JavaScript involved? There may be an issue with how sprockets is informed of dependencies. Are you seeing it with files being pulled in from npm or from local files you are editing?

**Comment by cymen** [comment]
@gauravtiwari Can you confirm you are using 2.0.2? Both 2.0.0 and 2.0.1 have the bug.

**Comment by gauravtiwari** [comment]
@cymen Let me quickly check and report - I have just checked this with with ```2.0.1``` and ```2.0.2```. I am using ```react-rails``` gem and basically whenever I make changes to components file it doesn't detect the changes, however if I make a change in ```application.js``` or ```components.js``` (under main ```app/javascripts``` tree) it works fine.

**Comment by gauravtiwari** [comment]
@cymen just to confirm, both ```2.0.2``` and ```2.0.1``` have this issue, it seems they are not detecting changes inside sub-directories.

**Comment by cymen** [comment]
@gauravtiwari Would you be able to push up a simple rails project using `react-rails` and `browserify-rails` that has the bug? I haven't used `react-rails` yet.

**Comment by gauravtiwari** [comment]
@cymen Here you go - [https://github.com/gauravtiwari/relay-rails-blog](https://github.com/gauravtiwari/relay-rails-blog) This is still in development, but if you run the app and try making any changes in ```javascripts/components``` you will see the browserify doesn't detect the changes.

**Comment by thesteady** [comment]
+1 on not detecting nested file changes still being an issue when trying to use react-rails with browserify-rails 2.0.2. Seems to work ok on 1.5.0!

**Comment by cymen** [comment]
I think the problem is react-rails. I'm kind of surprised anyone is using that with browserify-rails because it looks like a way to use react on the asset pipeline but not using CommonJS whereas browserify-rails is all about CommonJS. So if you wanted to use react with browserify-rails, you would go down the route of using something like reactify or another transform instead of a gem like react-rails. Setting that up can be a hassle but my motivation in working on browserify-rails is to push rails towards CommonJS so something like react-rails is a step backward from my perspective.

**Comment by thesteady** [comment]
@cymen, I agree with you. We started using react-rails as a quick way to start adding react to our app, but I've quickly begun hitting a lot of conflicts with the _"rails way"_ of trying to do javascript versus the _javascript_ way of doing javascript.

**Comment by cymen** [comment]
@thesteady Exactly. I hesitated to bring this up as I do think people should be free to use things as they see fit. But I think the happy path here is to not use react-rails.

But I think we need documentation in the README on how to use it with React as the configuration can be difficult.

**Comment by shaimo** [comment]
+1 on seeing this issue (I guess a similar setup with react-rails and browserify-rails). Every time I change some js code I need to clear the cache...:-(

**Comment by ShallmentMo** [comment]
same issue +1 (Also using react-rails). My version is 2.0.2

**Comment by gucki** [comment]
Same issue, but I'm not using react-rails gem.

**Comment by cymen** [comment]
Thanks for all the reports. Please keep making them. For those that are feeling blocked by this, please use the 1.5.x versions. There is little difference between the two (although in theory, 2.x should be a little more efficient). It may take some time to fix these issues so I don't want anyone to feel blocked from progress or wait for a fix when it may take a bit longer than we'd all like (PRs welcome though!).

@gucki Thanks -- it was unclear if it was just an issue w/ react-rails. Are you using any other gems that hook into the asset pipeline?

**Comment by oelmekki** [comment]
> I'm kind of surprised anyone is using that with browserify-rails because it looks like a way to use react on the asset pipeline but not using CommonJS whereas browserify-rails is all about CommonJS.

@cymen Just so you know, there is a good reason to want both browserify 
and react-rails: using browserify in general development while still 
being able to easily do server side rendering of react components ;)

This works great btw, except for this cache problem. I can confirm 
downgrading to 1.5.0 for now fixes it.

**Comment by shaimo** [comment]
@oelmekki Exactly! :-)

**Comment by adsteel** [comment]
Same issue here, and not using the React-Rails gems.

**Comment by sunflat** [comment]
I had the same issue, and made PR #130

**Comment by cymen** [comment]
@oelmekki Ah! Thanks for explaining. I haven't used react-rails and am now doing server-side rendering with a NodeJS engine but I can see how it would be very useful in the Rails environment.

@sunflat Thanks -- I'll test, merge and push and hopefully everyone else can pull the new version and test to verify. Thanks again.

**Comment by cymen** [comment]
I merged @sunflat's PR #130 and released it as browserify-rails 2.1.0. Would everyone who has time pull it down and try it? Thanks for your patience and thanks to @sunflat for the PR.

**Comment by adsteel** [comment]
@cymen I pulled and ran, things seem to be working. No more caching issues (https://github.com/browserify-rails/browserify-rails/issues/101) for us at least.

**Comment by sunflat** [comment]
I found that PR #130 doesn't work on some cases (when the file name doesn't change but its contents change).
So I made a new PR #131.

**Comment by cymen** [comment]
@sunflat Thanks -- that makes sense to me.

I've released 2.2.0 with PR #131.

**Comment by onebree** [comment]
Since PR #131 was merged, can this issue be closed?

**Comment by cymen** [comment]
Good point. Let's close it. Reopen with reason if we got it wrong somehow.
